File: Editorial/170_Two_Sum_III_Data_structure_design/CSharp/1_Sorted_List.cs
public class TwoSum {
    private List<int> nums;
    private bool is_sorted;

    /** Initialize your data structure here. */
    public TwoSum() {
        this.nums = new List<int>();
        this.is_sorted = false;
    }

    /** Add the number to an internal data structure.. */
    public void Add(int number) {
        this.nums.Add(number);
        this.is_sorted = false;
    }

    /** Find if there exists any pair of numbers which sum is equal to the value. */
    public bool Find(int value) {
        if (!this.is_sorted) {
            this.nums.Sort();
            this.is_sorted = true;
        }

        int low = 0, high = this.nums.Count - 1;
        while (low < high) {
            int twosum = this.nums[low] + this.nums[high];
            if (twosum < value)
                low += 1;
            else if (twosum > value)
                high -= 1;
            else
                return true;
        }

        return false;
    }
}

File: Editorial/170_Two_Sum_III_Data_structure_design/CPP/1_Sorted_List.cpp
class TwoSum {
private:
    std::vector<int> nums;
    bool is_sorted;

public:
    TwoSum() : is_sorted(false) {}

    void add(int number) {
        nums.push_back(number);
        is_sorted = false;
    }

    bool find(int value) {
        if (!is_sorted) {
            std::sort(nums.begin(), nums.end());
            is_sorted = true;
        }

        int low = 0;
        int high = nums.size() - 1;

        while (low < high) {
            int sum = nums[low] + nums[high];

            if (sum < value) {
                low += 1;
            } else if (sum > value) {
                high -= 1;
            } else {
                return true;
            }
        }

        return false;
    }
};
